src/serac/physics/utilities/functional/integral.hpp

```diff
@@ -439,6 +439,243 @@ void gradient_kernel(const mfem::Vector& dU, mfem::Vector& dR, derivatives_type*
+      auto                    J_q = make_tensor<spatial_dim, geometry_dim>([&](int i, int j) { return J(q, i, j, e); });
+      auto                    detJ_q = detail::Measure(J_q);
+      [[maybe_unused]] double dx     = detJ_q * dxi_q;
+
+      // recall the derivative of the q-function w.r.t. its arguments at this quadrature point
+      auto dq_darg = derivatives_ptr[e * int(rule.size()) + q];
+
+      // use the chain rule to compute the first-order change in the q-function output
+      // auto dq = chain_rule(dq_darg, darg);
+      [[maybe_unused]] auto M = test_element::shape_functions(xi_q);
+      [[maybe_unused]] auto N = trial_element::shape_functions(xi_q);
+      if constexpr (test_element::family == Family::HCURL) {
+        M = dot(M, inv(J_q));
+      }
+      if constexpr (trial_element::family == Family::HCURL) {
+        N = dot(N, inv(J_q));
+      }
+
+      auto f00 = std::get<0>(std::get<0>(dq_darg));
+      auto f01 = std::get<1>(std::get<0>(dq_darg));
+      auto f10 = std::get<0>(std::get<1>(dq_darg));
+      auto f11 = std::get<1>(std::get<1>(dq_darg));
+
+      // df0_du stiffness contribution
+      // size(M) = test_ndof
+      // size(N) = trial_ndof
+      // size(df0_du) = test_dim x trial_dim
+      if constexpr (!is_zero<decltype(f00)>::value) {
+        auto df0_du = convert_to_tensor_with_shape<test_dim, trial_dim>(f00);
+        for_loop<test_ndof, test_dim, trial_ndof, trial_dim>([&](auto i, auto id, auto j, auto jd) {
+          // maybe we should have a mapping for dofs x dim
+          K_elem[i + test_ndof * id][j + trial_ndof * jd] += M[i] * df0_du[id][jd] * N[j] * dx;
+        });
+      }
+
+      if constexpr (test_element::family == Family::H1 || test_element::family == Family::L2) {
+        [[maybe_unused]] auto dM_dx = dot(test_element::shape_function_gradients(xi_q), inv(J_q));
+        [[maybe_unused]] auto dN_dx = dot(trial_element::shape_function_gradients(xi_q), inv(J_q));
+
+        // df0_dgradu stiffness contribution
+        // size(M) = test_ndof
+        // size(df0_dgradu) = test_dim x trial_dim x spatial_dim
+        // size(dN_dx) = trial_ndof x spatial_dim
+        if constexpr (!is_zero<decltype(f01)>::value) {
+          auto df0_dgradu = convert_to_tensor_with_shape<test_dim, trial_dim, spatial_dim>(f01);
+          for_loop<test_ndof, test_dim, trial_ndof, trial_dim, spatial_dim>(
+              [&](auto i, auto id, auto j, auto jd, auto dummy_i) {
+                // maybe we should have a mapping for dofs x dim
+                K_elem[i + test_ndof * id][j + trial_ndof * jd] +=
+                    M[i] * df0_dgradu[id][jd][dummy_i] * dN_dx[j][dummy_i] * dx;
+              });
+        }
+
+        // // df1_du stiffness contribution
+        // size(dM_dx) = test_ndof x spatial_dim
+        // size(N) = trial_ndof
+        // size(df1_du) = test_dim x spatial_dim x trial_dim
+        if constexpr (!is_zero<decltype(f10)>::value) {
+          auto& df1_du = f10;
```

Comment:
some of these are just references (i.e. f10, f11), others convert to tensors with some assumed shape (i.e. f00, f01).

This part of the code will likely be refactored in the near future, but I was wondering if there was a reason for the separate treatment of those different derivative terms.

Reply:
No. The conversion routines are really workarounds to handle the derivatives being `serac::zero` or `double` types. I tried to keep all my matrix code such that it uses tensor access patterns.

The cases that are not "converted" may also need to undergo "conversion". I don't think they have been well vetted by the functional_comparison tests.